portfolio/src/main/java/com/google/sps/servlets/HomeServlet.java

```diff
@@ -1,38 +1,86 @@
-      String urlToRedirectToAfterUserLogsOut = HOME_PATH;
-      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
 
-      response.getWriter().println("<p>Hello " + currentUserEmail + "!</p>");
-      response.getWriter().println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+
+    if (!userService.isUserLoggedIn()) {
+      String loginUrl = userService.createLoginURL(HOME_PATH);
+
+      response.getWriter().println("<style>#comment-form {display:none;}</style>");
+      response.getWriter().println("<p>Hello friend.</p>");
+      response.getWriter().println("<p>Please log in to Add a comment.</p>");
+      response.getWriter().println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
       return;
     }
 
-    String urlToRedirectToAfterUserLogsIn = HOME_PATH;
-    String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+    // If user has not set a nickname, redirect to nickname page
+    Optional<String> optionalNickname = getUserNickname(userService.getCurrentUser().getUserId());
+    if (!optionalNickname.isPresent()) {
+      response.sendRedirect(NICKNAME_SERVLET);
+      return;    
+    }
+    
+    nickname = optionalNickname.get();
+    userEmail = userService.getCurrentUser().getEmail();
+    String logoutUrl = userService.createLogoutURL(HOME_PATH);
+
+    response.getWriter().println("<p>Hello " + nickname + "!</p>");
+    response.getWriter().println("<button onclick=\"changeNickname();\">Change nickname</button>");
+    
+    // Allows the user to delete all of the comments only if the email is the same as the admin email.
+    if (userEmail.equals(ADMIN_EMAIL)) {
+      response.getWriter().println("<button onclick=\"deleteAllComments();\">Delete all Comments</button>");
+    }
+
+    response.getWriter().println("<p>Logout <a href=\"" + logoutUrl + "\">here</a>.</p>");
+  }
+
+  /** Returns an Optional object containing the nickname of the user with {@code id},
+   * or an empty Optional if the user has not set a nickname. */
+  private Optional<String> getUserNickname(String id) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query =
+        new Query(USER_INFO)
+            .setFilter(new Query.FilterPredicate(ID, Query.FilterOperator.EQUAL, id));
+    PreparedQuery results = datastore.prepare(query);
+    Optional<Entity> optionalEntity = Optional.ofNullable(results.asSingleEntity());
+    
+    return optionalEntity.map(entity->(String)entity.getProperty(COMMENT_NICKNAME));
+  }
 
-    response.getWriter().println("<style>form {display:none;}</style>");
-    response.getWriter().println("<p>Hello friend.</p>");
-    response.getWriter().println("<p>Please log in to Add a comment.</p>");
-    response.getWriter().println("<p>Login <a href=\"" + loginUrl + "\">here</a>.</p>");
+  /** Returns the nickname of the current user when needed in other classes. */
+  public static String getUserNickname() {
```

Comment:
This is bug prone. If doGet was not called before getUserNickname then this will return a null string.

Reply:
With my implementation, getUserNickname() is only called after the user submits a nickname so it should never be null.  I ensure this is the case by preventing a user from submitting a comment until they enter a nickname.

I updated my method header to mention this. Should I rewrite my method to use Optional?